syntax = "proto3";

service Deliverer {
	
	//1)
	//operação sem parâmetros e sem valor de retorno
	rpc void_function (Empty) returns (Empty) {}

	//2)
	//operação com um long de parâmetro e retorna um long
	rpc duplicate_function(SumLongRequest) returns (SumLongResponse){}

	//3)
	//operação com um 8 valores long de parâmetro e retorna um long
	rpc adder_function(SumEightLongsRequest) returns (SumLongResponse){}

	//4)
	//operação com parâmetro string e valor de retorno string
	rpc maximizeString_function(StringRequest) returns (StringResponse){}

	//5) Operação que passa um tipo complexo e retorna outro tipo complexo
	rpc criaPessoa(PessoaEndereco) returns (PessoaResposta){}

	// Insere um endereço no banco de dados e retorna seu id e uma data
	rpc adicionaEndereco(Endereco) returns (PessoaResposta){}

	// Passa um retângulo e retorna outro retângulo com o dobro de area
	rpc esticaRetangulo(RetanguloRequest) returns (RetanguloRequest){}
}

message Empty{}

message SumLongRequest {
	int64 a = 1;
}

message SumLongResponse {
	int64 result = 1;
}

message SumEightLongsRequest {
	int64 a = 1;
	int64 b = 2;
	int64 c = 3;
	int64 d = 4;
	int64 e = 5;
	int64 f = 6;
	int64 g = 7;
	int64 h = 8;
}

message StringRequest{
	string a = 1;
}

message StringResponse{
	string result = 1;
}

message PessoaEndereco{
	Pessoa person = 1;
	Endereco address = 2;
}

message Endereco{
	string cidade = 1;
	string rua = 2;
	int32 numero = 3;
}

message Pessoa{
	string nome = 1;
	int32 idade = 2;
}

message PessoaResposta{
	int32 id_pessoa = 1;
	string hora = 2;
}

message RetanguloRequest{
	Retangulo ret = 1;
}

message Retangulo{
	int32 base = 1;
	int32 altura = 2;
}